### PR TITLE
[SPARK-49596][SQL] Improve performance of `FormatString`

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to improve performance of the expression `FormatString`.

### Why are the changes needed?
- In the expression `FormatString`, the class `java.lang.StringBuffer` used is `thread safe`, but its performance is not as good as `java.lang.StringBuilder`.
- In the scenario we are using, there is actually no need to use `java.lang.StringBuffer`.
- In class `java.util.Formatter`, it also defaults to using `java.lang.StringBuilder`, as follows:
  image

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA.
Manually check.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by panbingkun** [comment]
- Benchmark code as follows:
```scala
object PrintfBenchmark extends SqlBasedBenchmark {
  private val N = 10_000_00
  private val M = 100

  private val df = spark.range(N).to(new StructType().add("id", "int")).
    withColumn("format", lit("aa%d%s")).
    withColumn("value1", col("id") % M).
    withColumn("value2", lit("xxx"))

  private def doBenchmark(): Unit = {
    df.selectExpr("printf(format, value1, value2)").noop()
  }

  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
    runBenchmark("printf") {
      val benchmark = new Benchmark("printf", N, output = output)
      benchmark.addCase("optimize", M) { _ =>
        doBenchmark()
      }
      benchmark.run()
    }
  }
}
```

- Before (run 3 times)
```

Running benchmark: printf
  Running case: has foldable optimize
  Stopped after 100 iterations, 25558 ms

OpenJDK 64-Bit Server VM 17.0.10+7-LTS on Mac OS X 14.6.1
Apple M2
printf:                                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
optimize                                            246            256          21          4.1         245.8       1.0X


Running benchmark: printf
  Running case: has foldable optimize
  Stopped after 100 iterations, 24545 ms

OpenJDK 64-Bit Server VM 17.0.10+7-LTS on Mac OS X 14.6.1
Apple M2
printf:                                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
optimize                                            241            245           5          4.2         240.6       1.0X

Running benchmark: printf
  Running case: has foldable optimize
  Stopped after 100 iterations, 26481 ms

OpenJDK 64-Bit Server VM 17.0.10+7-LTS on Mac OS X 14.6.1
Apple M2
printf:                                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
optimize                                            257            265          16          3.9         256.6       1.0X
```

- After (run 3 times)
```
Running benchmark: printf
  Running case: has foldable optimize
  Stopped after 100 iterations, 22811 ms

OpenJDK 64-Bit Server VM 17.0.10+7-LTS on Mac OS X 14.6.1
Apple M2
printf:                                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
optimize                                            217            228          39          4.6         216.7       1.0X


Running benchmark: printf
  Running case: has foldable optimize
  Stopped after 100 iterations, 21136 ms

OpenJDK 64-Bit Server VM 17.0.10+7-LTS on Mac OS X 14.6.1
Apple M2
printf:                                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
optimize                                            206            211           6          4.9         205.7       1.0X


Running benchmark: printf
  Running case: has foldable optimize
  Stopped after 100 iterations, 21805 ms

OpenJDK 64-Bit Server VM 17.0.10+7-LTS on Mac OS X 14.6.1
Apple M2
printf:                                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
optimize                                            211            218          16          4.7         210.8       1.0X
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 1]
```diff
@@ -1969,13 +1969,10 @@ case class FormatString(children: Expression*) extends Expression with ImplicitC
...
-      val sb = new StringBuffer()
-      val formatter = new java.util.Formatter(sb, Locale.US)
-
+      val formatter = new java.util.Formatter(Locale.US)
       val arglist = children.tail.map(_.eval(input).asInstanceOf[AnyRef])
-      formatter.format(pattern.asInstanceOf[UTF8String].toString, arglist: _*)
-
-      UTF8String.fromString(sb.toString)
+      UTF8String.fromString(
+        formatter.format(pattern.asInstanceOf[UTF8String].toString, arglist: _*).toString)
```
Comment (panbingkun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:1975]: image

**Comment by panbingkun** [comment]
cc @MaxGekk @cloud-fan

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @panbingkun.

**Comment by cloud-fan** [comment]
late LGTM

**Pull request closed without merging**